fix(ngCsp): allow CSP to be configurable

petebacondarwin opened the pull request:
There are two different features in Angular that can break CSP rules:
use of `eval` to execute a string as JavaScript and dynamic injection of
CSS style rules into the DOM.

This change allows us to configure which of these features should be turned
off to allow a more fine grained set of CSP rules to be supported.

Closes #11933
Closes #8459

petebacondarwin committed:
fix(ngCsp): allow CSP to be configurable

There are two different features in Angular that can break CSP rules:
use of `eval` to execute a string as JavaScript and dynamic injection of
CSS style rules into the DOM.

This change allows us to configure which of these features should be turned
off to allow a more fine grained set of CSP rules to be supported.

Closes #11933
Closes #8459

petebacondarwin commented:
@jdalton, @realityking, @lgalfaso and @btford - can you take a look?

jdalton commented on src/Angular.js:
```diff
@@ -984,22 +984,39 @@ function equals(o1, o2) {
+      csp.rules = {
+        noUnsafeEval: !ngCspAttribute || (ngCspAttribute.indexOf('no-unsafe-eval') !== -1),
+        noInlineStyle: !ngCspAttribute || (ngCspAttribute.indexOf('no-inline-style') !== -1)
+      };
```
Shouldn't the `ngCspAttribute.indexOf` checks be looking for `> -1` as if they're declared then they're `true`?

realityking replied:
Checking for `!== -1` has the same effect, so this is fine.

jdalton replied:
> Checking for `!== -1` has the same effect, so this is fine.

Doh! I blame it on all the double negatives :smile_cat:

jdalton commented on src/ng/directive/ngCsp.js:
```diff
@@ -35,11 +37,37 @@
+ * run eval - no automcatic check for unsafe eval will occur. E.g. `<body ng-csp="no-inline-style">`
+ *
+ * * Specifying both `no-unsafe-eval` and `no-inline-style` tells Angular that we must not inject
+ * styles nor use eval, which is the same as an empty: ng-csp, except that no runtime check for
```
the comment
> except that no runtime check for unsafe eval will occur.

can be removed as the runtime check won't occur for `ng-csp` either at the moment.

petebacondarwin replied:
Oh yes you are right.

realityking commented on src/ng/parse.js:
```diff
@@ -1703,11 +1703,11 @@ function $ParseProvider() {
   this.$get = ['$filter', '$sniffer', function($filter, $sniffer) {
     var $parseOptions = {
-          csp: $sniffer.csp,
+          csp: $sniffer.csp && $sniffer.csp.noUnsafeEval,
```
Looking at [$sniffer](https://github.com/petebacondarwin/angular.js/blob/ae2241522611cebc8d0833f0bed58bb726c6fdb1/src/ng/sniffer.js#L81), I think the check if  `$sniffer.csp ` exists is unnecessary.

petebacondarwin replied:
It would be except that there are loads of tests that mock out sniffer and don't provide a `csp` object. I debated back and forth between fixing the tests or fixing this line and in the end I felt that this was simpler as it meant all these other tests didn't need to know about the csp object.

You can see this in action if you remove this check and then run `grunt autotest`

realityking replied:
Ah good point. Probably an indication that `$sniffer`is doing too much, but that ship has sailed.

petebacondarwin replied:
Actually, perhaps we don't need to use `$sniffer` here at all? Since we are inside the angular.js closure we could just use the `csp()` function. I don't think we are mocking it anywhere are we?

realityking commented:
@petebacondarwin I like it (a lot)!

gkalpak commented on test/AngularSpec.js:
```diff
@@ -793,36 +804,51 @@ describe('angular', function() {
+    it('should return true when CSP is enabled manually via [data-ng-csp]', function() {
+      var spy = mockCspElement('data-ng-csp');
+      expect(csp()).toEqual({ noUnsafeEval: true, noInlineStyle: true });
       expect(document.querySelector).toHaveBeenCalledWith('[data-ng-csp]');
```
`expect(spy)` for consistency ? :smile:

gkalpak replied:
`expect(spy)` for consistency ? :smile:

petebacondarwin replied:
Right!

gkalpak commented on test/AngularSpec.js:
```diff
@@ -793,36 +804,51 @@ describe('angular', function() {
+    it('should return true for noUnsafeEval if it is specified in the `ng-csp` attribute value', function() {
+      var spy = mockCspElement('ng-csp', 'no-unsafe-eval');
+      expect(csp()).toEqual({ noUnsafeEval: true, noInlineStyle: false });
+    });
```
It might be also nice to test that no check is performed (i.e. spying on `window.Function` to verify it hasn't been called or something).

petebacondarwin replied:
Right, too

petebacondarwin committed:
fix(ngCsp): allow CSP to be configurable

There are two different features in Angular that can break CSP rules:
use of `eval` to execute a string as JavaScript and dynamic injection of
CSS style rules into the DOM.

This change allows us to configure which of these features should be turned
off to allow a more fine grained set of CSP rules to be supported.

Closes #11933
Closes #8459

petebacondarwin commented:
I have added a commit with changes thanks to the feedback.

gkalpak commented:
LGTM :+1:

btford commented:
:+1:
